single_include/nlohmann/json.hpp

```diff
@@ -4694,6 +4694,10 @@ template<typename BasicJsonType, typename CompatibleString,
          enable_if_t<std::is_constructible<typename BasicJsonType::string_t, CompatibleString>::value, int> = 0>
 void to_json(BasicJsonType& j, const CompatibleString& s)
 {
+    if (!j.is_null())
+    {
+        j = BasicJsonType();
+    }
```

Comment:
I think this needs to be in *every* `to_json` function that calls `external_constructor`.

Reply:
Yes, this was just one fix to fix the added test. If we agree this is the way to go, I will add more tests and fixes.

Reply:
If this is the most efficient way to properly clear out the existing object, then that looks good.  

I was thinking that a `clear()` member would be useful there, but we already have that and it doesn't delete the underlying thing that we're currently leaking.  A `set_to_null()` member would be useful here just to avoid having to retype that same block every time, and also to avoid the work of creating and destroying the temporary.

    void set_to_null()
    {
        if (!is_null())
        {
            m_value.destroy(m_type);
            m_type = value_t::null;
            m_value = nullptr;
         }
    }

On the other hand, I'm not sure this belongs in the public API and while `external_constructor` is a friend, `to_json` is not, so maybe this is better:

    namespace nlohmann
    {
    namespace detail
    {
        void set_to_null(BasicJsonType &j)
        {
            if (!j.is_null())
            {
                j = BasicJsonType{};
            }
        }
    }
    }


Hm. `external_constructor` is only used in `to_json`, the thing I found in another spot that called `::construct(` wasn't using `external_constructor`, so it is possible to put this in `external_constructor<>::create` without adding extra overhead, and since this is a friend, we can use an internal helper.

Reply:
You mean `external_constructor<>::construct` right?

Reply:
Yes, typing from memory, will fix.  :)

Reply:
As `external_constructor` is a friend, all we need is to add line

```cpp
j.m_value.destroy(j.m_type);
```

to all `external_constructor<>::construct` functions.

Reply:
Excellent.